op-supervisor/supervisor/backend/safety/safety.go

```diff
@@ -0,0 +1,274 @@
+		if err != nil {
+			return fmt.Errorf("failed to open iterator for chain %s block %d", chainID, ref.Number)
+		}
+		view = &View{
+			chainID: chainID,
+			iter:    iter,
+			localView: heads.HeadPointer{
+				LastSealedBlockHash: ref.Hash,
+				LastSealedBlockNum:  ref.Number,
+				LastSealedTimestamp: ref.Time,
+				LogsSince:           0,
+			},
+			localDerivedFrom: eth.L1BlockRef{},
+			validWithinView:  r.ValidWithinUnsafeView,
+		}
+		r.unsafe[chainID] = view
+	} else if err := view.UpdateLocal(eth.L1BlockRef{}, ref); err != nil {
+		return fmt.Errorf("failed to update local-unsafe: %w", err)
+	}
+	local, _ := r.unsafe[chainID].Local()
+	r.log.Debug("Updated local unsafe head", "chainID", chainID, "local", local)
+	r.advanceCrossUnsafe()
+	return nil
+}
+
+// advanceCrossUnsafe calls Process on all cross-unsafe views.
+func (r *safetyIndex) advanceCrossUnsafe() {
+	for chainID, view := range r.unsafe {
+		if err := view.Process(); err != nil {
+			r.log.Error("Failed to update cross-unsafe view", "chain", chainID, "err", err)
+		}
+		cross, _ := r.unsafe[chainID].Cross()
+		r.log.Debug("Updated cross unsafe head", "chainID", chainID, "cross", cross)
+	}
+}
+
+// UpdateLocalSafe updates the local-safe view for the given chain, and advances the cross-safe status.
+func (r *safetyIndex) UpdateLocalSafe(
+	chainID types.ChainID, at eth.L1BlockRef, ref eth.L2BlockRef) error {
+	view, ok := r.safe[chainID]
+	if !ok {
+		iter, err := r.chains.IteratorStartingAt(chainID, ref.Number, 0)
+		if err != nil {
+			return fmt.Errorf("failed to open iterator for chain %s block %d", chainID, ref.Number)
+		}
+		view = &View{
+			chainID: chainID,
+			iter:    iter,
+			localView: heads.HeadPointer{
+				LastSealedBlockHash: ref.Hash,
+				LastSealedBlockNum:  ref.Number,
+				LastSealedTimestamp: ref.Time,
+				LogsSince:           0,
+			},
+			localDerivedFrom: at,
+			validWithinView:  r.ValidWithinSafeView,
+		}
+		r.safe[chainID] = view
+	} else if err := view.UpdateLocal(at, ref); err != nil {
+		return fmt.Errorf("failed to update local-safe: %w", err)
+	}
+
+	// register what this L2 block is derived from
+	m, ok := r.derivedFrom[chainID]
+	if !ok {
+		m = make(map[common.Hash]eth.L1BlockRef)
+		r.derivedFrom[chainID] = m
+	}
+	m[ref.Hash] = at
+	local, _ := r.safe[chainID].Local()
+	r.log.Debug("Updated local safe head", "chainID", chainID, "local", local)
+	r.advanceCrossSafe()
+	return nil
+}
+
+// advanceCrossSafe calls Process on all cross-safe views, and advances the finalized safety status.
+func (r *safetyIndex) advanceCrossSafe() {
+	for chainID, view := range r.safe {
+		if err := view.Process(); err != nil {
+			r.log.Error("Failed to update cross-safe view", "chain", chainID, "err", err)
+		}
+		cross, _ := r.safe[chainID].Cross()
+		r.log.Debug("Updated local safe head", "chainID", chainID, "cross", cross)
+	}
+	r.advanceFinalized()
+	// TODO prune any L2 derivedFrom entry older than the L2 finalized entry
```

Comment:
```suggestion
```
Once we're not doing the safety index fully in memory, then there's no need for pruning anymore. Having a complete historical safety-index view is actually useful. At least also to cover the span of duration for fault-proof games to determine accurate L1-input / L2-anchor context.